File: src/omnia/data/structure/observable/writable/WritableObservableDirectedGraphImpl.java
package omnia.data.structure.observable.writable;

import static java.util.Objects.requireNonNull;
import static omnia.data.stream.Collectors.toImmutableSet;
import static omnia.data.stream.Collectors.toSet;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.processors.PublishProcessor;
import java.util.Optional;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;
import omnia.data.stream.Streams;
import omnia.data.structure.DirectedGraph;
import omnia.data.structure.Set;
import omnia.data.structure.immutable.ImmutableDirectedGraph;
import omnia.data.structure.immutable.ImmutableSet;
import omnia.data.structure.observable.ObservableDirectedGraph;
import omnia.data.structure.tuple.Couplet;

final class WritableObservableDirectedGraphImpl<E> implements WritableObservableDirectedGraph<E> {

  private volatile ImmutableDirectedGraph<E> state;
  private final FlowableProcessor<MutationEvent<E>> mutationEventProcessor =
      PublishProcessor.<MutationEvent<E>>create().toSerialized();

  WritableObservableDirectedGraphImpl() {
    state = ImmutableDirectedGraph.empty();
  }

  WritableObservableDirectedGraphImpl(DirectedGraph<E> original) {
    state = ImmutableDirectedGraph.copyOf(original);
  }

  @Override
  public void addNode(E item) {
    mutateState(
        currentState -> !currentState.contents().contains(item),
        currentState -> currentState.toBuilder().addNode(item).build(),
        (previousState, newState) -> ImmutableSet.of(AddNodeToGraph.create(item)));
  }

  @Override
  public void replaceNode(E original, E replacement) {
    mutateState(
        currentState -> {
          if (!currentState.contents().contains(original)) {
            throw new IllegalArgumentException(
                "cannot replace a non-existent node. original=" + original);
          }
          if (currentState.contents().contains(replacement)) {
            throw new IllegalArgumentException(
                "cannot replace a node with an already existing node. replacement=" + replacement);
          }
          return true;
        },
        currentState -> currentState.toBuilder().replaceNode(original, replacement).build(),
        (previousState, newState) ->
            Streams.concat(
                previousState.nodeOf(original)
                    .stream()
                    .map(DirectedNode::edges)
                    .flatMap(Set::stream)
                    .map(DirectedEdge::endpoints)
                    .map(couplet -> couplet.map(DirectedNode::item))
                    .map(RemoveEdgeFromGraph::create),
                previousState.nodeOf(original)
                    .stream()
                    .map(DirectedNode::item)
                    .map(RemoveNodeFromGraph::create),
                newState.nodeOf(replacement)
                    .stream()
                    .map(DirectedNode::item)
                    .map(AddNodeToGraph::create),
                newState.nodeOf(replacement)
                    .stream()
                    .map(DirectedNode::edges)
                    .flatMap(Set::stream)
                    .map(DirectedEdge::endpoints)
                    .map(couplet -> couplet.map(DirectedNode::item))
                    .map(AddEdgeToGraph::create))
                .collect(toSet()));
  }

  @Override
  public boolean removeUnknownTypedNode(Object item) {
    return mutateState(
        currentState -> currentState.contents().containsUnknownTyped(item),
        currentState -> currentState.toBuilder().removeUnknownTypedNode(item).build(),
        (previousState, newState) ->
            Streams.concat(
                previousState.nodeOf(item).stream()
                    .map(DirectedNode::edges)
                    .flatMap(Set::stream)
                    .map(DirectedEdge::endpoints)
                    .map(couplet -> couplet.map(DirectedNode::item))
                    .map(RemoveEdgeFromGraph::create),
                previousState.nodeOf(item).stream()
                    .map(DirectedNode::item)
                    .map(RemoveNodeFromGraph::create))
                .collect(toSet()));
  }

  @Override
  public void addEdge(E from, E to) {
    mutateState(
        currentState -> currentState.edgeOf(from, to).isEmpty(),
        currentState -> currentState.toBuilder().addEdge(from, to).build(),
        (previousState, newState) ->
            newState.edgeOf(from, to).stream()
                .map(DirectedEdge::endpoints)
                .map(couplet -> couplet.map(DirectedNode::item))
                .map(AddEdgeToGraph::create)
                .collect(toSet()));
  }

  @Override
  public boolean removeEdge(E from, E to) {
    return removeEdgeUnknownTyped(from, to);
  }

  @Override
  public boolean removeEdgeUnknownTyped(Object from, Object to) {
    return mutateState(
        currentState -> currentState.edgeOf(from, to).isPresent(),
        currentState -> currentState.toBuilder().removeEdgeUnknownEdge(from, to).build(),
        (previousState, newState) ->
            previousState.edgeOf(from, to).stream()
                .map(DirectedEdge::endpoints)
                .map(couplet -> couplet.map(DirectedNode::item))
                .map(RemoveEdgeFromGraph::create)
                .collect(toSet()));
  }

  private boolean mutateState(
      Predicate<? super ImmutableDirectedGraph<E>> shouldChange,
      Function<? super ImmutableDirectedGraph<E>, ? extends ImmutableDirectedGraph<E>> mutateState,
      BiFunction<? super ImmutableDirectedGraph<E>, ? super ImmutableDirectedGraph<E>, ? extends Set<GraphOperation<E>>> mutationOperations) {
    ImmutableDirectedGraph<E> previousState;
    ImmutableDirectedGraph<E> nextState;
    synchronized (this) {
      previousState = state;
      if (!shouldChange.test(previousState)) {
        return false;
      }
      nextState = requireNonNull(mutateState.apply(previousState));
      state = nextState;
      Set<GraphOperation<E>> operations = mutationOperations.apply(previousState, nextState);
      mutationEventProcessor.onNext(
          new MutationEvent<>() {
            @Override
            public DirectedGraph<E> state() {
              return nextState;
            }

            @Override
            public Set<GraphOperation<E>> operations() {
              return operations;
            }
          });
    }
    return true;
  }

  @Override
  public Optional<? extends DirectedNode<E>> nodeOf(Object item) {
    return getState().nodeOf(item);
  }

  @Override
  public Optional<? extends DirectedEdge<E>> edgeOf(Object from, Object to) {
    return getState().edgeOf(from, to);
  }

  @Override
  public Set<E> contents() {
    return getState().contents();
  }

  @Override
  public Set<? extends DirectedNode<E>> nodes() {
    return getState().nodes();
  }

  @Override
  public Set<? extends DirectedEdge<E>> edges() {
    return getState().edges();
  }

  @Override
  public ObservableDirectedGraph<E> toReadOnly() {
    return new ObservableDirectedGraph<>() {
      @Override
      public ObservableChannels<E> observe() {
        return WritableObservableDirectedGraphImpl.this.observe();
      }

      @Override
      public Optional<? extends DirectedNode<E>> nodeOf(Object item) {
        return WritableObservableDirectedGraphImpl.this.nodeOf(item);
      }

      @Override
      public Optional<? extends DirectedEdge<E>> edgeOf(E from, E to) {
        return WritableObservableDirectedGraphImpl.this.edgeOf(from, to);
      }

      @Override
      public Set<? extends DirectedNode<E>> nodes() {
        return WritableObservableDirectedGraphImpl.this.nodes();
      }

      @Override
      public Set<? extends DirectedEdge<E>> edges() {
        return WritableObservableDirectedGraphImpl.this.edges();
      }

      @Override
      public Set<E> contents() {
        return WritableObservableDirectedGraphImpl.this.contents();
      }
    };
  }

  @Override
  public ObservableDirectedGraph.ObservableChannels<E> observe() {
    return new ObservableDirectedGraph.ObservableChannels<>() {
      @Override
      public Flowable<DirectedGraph<E>> states() {
        return Flowable.create(
            emitter -> {
              synchronized (this) {
                emitter.onNext(getState());
                emitter.setDisposable(
                    mutationEventProcessor.map(MutationEvent::state)
                        .subscribe(emitter::onNext, emitter::onError, emitter::onComplete));
              }
            },
            BackpressureStrategy.LATEST);
      }

      @Override
      public Flowable<MutationEvent<E>> mutations() {
        return Flowable.create(
            emitter -> {
               synchronized (this) {
                 ImmutableDirectedGraph<E> state = getState();
                 Set<GraphOperation<E>> operations =
                     Streams.concat(
                         state.nodes().stream()
                             .map(DirectedNode::item)
                             .map(AddNodeToGraph::create),
                         state.edges().stream()
                             .map(DirectedEdge::endpoints)
                             .map(couplet -> couplet.map(DirectedNode::item))
                             .map(AddEdgeToGraph::create))
                         .collect(toImmutableSet());
                 emitter.onNext(new MutationEvent<>() {
                   @Override
                   public DirectedGraph<E> state() {
                     return state;
                   }

                   @Override
                   public Set<GraphOperation<E>> operations() {
                     return operations;
                   }
                 });
                 emitter.setDisposable(
                     mutationEventProcessor
                         .subscribe(emitter::onNext, emitter::onError, emitter::onComplete));
               }
            },
            BackpressureStrategy.BUFFER);
      }
    };
  }

  private ImmutableDirectedGraph<E> getState() {
    synchronized (this) {
      return state;
    }
  }

  interface MutationEvent<E> extends ObservableDirectedGraph.MutationEvent<E> {

    @Override
    DirectedGraph<E> state();

    @Override
    Set<GraphOperation<E>> operations();
  }

  private static abstract class NodeOperation<E> implements ObservableDirectedGraph.NodeOperation<E> {

    private final E item;

    private NodeOperation(E item) {
      this.item = requireNonNull(item);
    }

    @Override
    public final E item() {
      return item;
    }
  }

  private static final class AddNodeToGraph<E> extends NodeOperation<E> implements ObservableDirectedGraph.AddNodeToGraph<E> {
    private AddNodeToGraph(E item) {
      super(item);
    }

    static <E> AddNodeToGraph<E> create(E item) {
      return new AddNodeToGraph<>(item);
    }
  }

  private static final class RemoveNodeFromGraph<E> extends NodeOperation<E> implements ObservableDirectedGraph.RemoveNodeFromGraph<E> {
    private RemoveNodeFromGraph(E item) {
      super(item);
    }

    static <E> RemoveNodeFromGraph<E> create(E item) {
      return new RemoveNodeFromGraph<>(item);
    }
  }

  private static abstract class EdgeOperation<E> implements ObservableDirectedGraph.EdgeOperation<E> {
    private final Couplet<E> endpoints;

    private EdgeOperation(Couplet<E> endpoints) {
      this.endpoints = requireNonNull(endpoints);
    }

    @Override
    public final Couplet<E> endpoints() {
      return endpoints;
    }
  }

  private static final class AddEdgeToGraph<E> extends EdgeOperation<E> implements ObservableDirectedGraph.AddEdgeToGraph<E> {
    private AddEdgeToGraph(Couplet<E> endpoints) {
      super(endpoints);
    }

    static <E> AddEdgeToGraph<E> create(Couplet<E> endpoints) {
      return new AddEdgeToGraph<>(endpoints);
    }
  }

  private static final class RemoveEdgeFromGraph<E> extends EdgeOperation<E> implements ObservableDirectedGraph.RemoveEdgeFromGraph<E> {
    private RemoveEdgeFromGraph(Couplet<E> endpoints) {
      super(endpoints);
    }

    static <E> RemoveEdgeFromGraph<E> create(Couplet<E> endpoints) {
      return new RemoveEdgeFromGraph<>(endpoints);
    }
  }
}

File: src/omnia/data/structure/observable/writable/WritableObservableDirectedGraphImpl.kt
package omnia.data.structure.observable.writable

import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.ObservableEmitter
import io.reactivex.rxjava3.subjects.PublishSubject
import java.util.Objects
import java.util.function.BiFunction
import java.util.function.Function
import java.util.function.Predicate
import java.util.stream.Stream
import omnia.data.stream.Collectors
import omnia.data.stream.Streams
import omnia.data.structure.DirectedGraph
import omnia.data.structure.Set
import omnia.data.structure.immutable.ImmutableDirectedGraph
import omnia.data.structure.immutable.ImmutableSet
import omnia.data.structure.observable.ObservableDirectedGraph
import omnia.data.structure.observable.ObservableGraph
import omnia.data.structure.observable.ObservableGraph.GraphOperation
import omnia.data.structure.tuple.Couplet

internal class WritableObservableDirectedGraphImpl<E : Any> : WritableObservableDirectedGraph<E> {

  @Volatile
  private var state: ImmutableDirectedGraph<E>
  private val mutationEvents = PublishSubject.create<MutationEvent<E>>().toSerialized()

  constructor() {
    state = ImmutableDirectedGraph.empty()
  }

  constructor(original: DirectedGraph<E>) {
    state = ImmutableDirectedGraph.copyOf(original)
  }

  override fun addNode(item: E) {
    mutateState(
      { currentState -> !currentState.contents().contains(item) },
      { currentState -> currentState.toBuilder().addNode(item).build() }
    ) { _, _ -> ImmutableSet.of(AddNodeToGraph.create(item)) }
  }

  override fun replaceNode(original: E, replacement: E) {
    mutateState(
      { currentState: ImmutableDirectedGraph<E> ->
        require(
          currentState.contents().contains(original)
        ) { "cannot replace a non-existent node. original=$original" }
        require(
          !currentState.contents().contains(replacement)
        ) { "cannot replace a node with an already existing node. replacement=$replacement" }
        true
      },
      { currentState -> currentState.toBuilder().replaceNode(original, replacement).build() }
    ) { previousState, newState ->
      Streams.concat(
          (previousState.nodeOf(original)?.edges() ?: ImmutableSet.empty())
              .stream()
              .map { it.endpoints() }
              .map { couplet -> couplet.map(Function { it.item() }) }
              .map { RemoveEdgeFromGraph.create(it) },
          previousState.nodeOf(original)
              ?.item()
              ?.let { RemoveNodeFromGraph.create(it) }
              ?.let { Stream.of(it) }
              ?: Stream.empty(),
          newState.nodeOf(replacement)
              ?.item()
              ?.let { AddNodeToGraph.create(it) }
              ?.let { Stream.of(it) }
              ?: Stream.empty(),
          (newState.nodeOf(replacement)?.edges() ?: ImmutableSet.empty())
              .stream()
              .map { it.endpoints() }
              .map { couplet -> couplet.map(Function { it.item() }) }
              .map { AddEdgeToGraph.create(it) })
        .collect(Collectors.toSet())
    }
  }

  override fun removeUnknownTypedNode(item: Any?): Boolean {
    return mutateState(
      { currentState -> currentState.contents().containsUnknownTyped(item) },
      { currentState -> currentState.toBuilder().removeUnknownTypedNode(item).build() }
    ) { previousState, _ ->
      Streams.concat(
          (previousState.nodeOfUnknownType(item)?.edges() ?: ImmutableSet.empty())
              .stream()
              .map { it.endpoints() }
              .map { couplet -> couplet.map(Function { it.item() }) }
              .map { RemoveEdgeFromGraph.create(it) },
          previousState.nodeOfUnknownType(item)
              ?.item()
              ?.let { RemoveNodeFromGraph.create(it) }
              ?.let { Stream.of(it) }
              ?: Stream.empty())
        .collect(Collectors.toSet())
    }
  }

  override fun addEdge(from: E, to: E) {
    mutateState(
      { currentState -> currentState.edgeOf(from, to) == null },
      { currentState -> currentState.toBuilder().addEdge(from, to).build() }
    ) { _, newState ->
      newState.edgeOf(from, to)
          ?.endpoints()
          ?.map(Function { it.item() })
          ?.let { AddEdgeToGraph.create(it) }
          ?.let { ImmutableSet.of(it) }
          ?: ImmutableSet.empty()
    }
  }

  override fun removeEdge(from: E, to: E): Boolean {
    return removeEdgeUnknownTyped(from, to)
  }

  override fun removeEdgeUnknownTyped(from: Any?, to: Any?): Boolean {
    return mutateState(
      { currentState -> currentState.edgeOfUnknownType(from, to) != null },
      { currentState -> currentState.toBuilder().removeEdgeUnknownEdge(from, to).build() }
    ) { previousState, _ ->
      previousState.edgeOfUnknownType(from, to)
          ?.endpoints()
          ?.map(Function { it.item() })
          ?.let { RemoveEdgeFromGraph.create(it) }
          ?.let { ImmutableSet.of(it) }
          ?: ImmutableSet.empty()
    }
  }

  private fun mutateState(
    shouldChange: Predicate<in ImmutableDirectedGraph<E>>,
    mutateState: Function<in ImmutableDirectedGraph<E>, out ImmutableDirectedGraph<E>>,
    mutationOperations: BiFunction<in ImmutableDirectedGraph<E>, in ImmutableDirectedGraph<E>, out Set<GraphOperation<E>>>
  ): Boolean {
    var previousState: ImmutableDirectedGraph<E>
    var nextState: ImmutableDirectedGraph<E>
    synchronized(this) {
      previousState = state
      if (!shouldChange.test(previousState)) {
        return false
      }
      nextState = Objects.requireNonNull(mutateState.apply(previousState))
      state = nextState
      val operations = mutationOperations.apply(previousState, nextState)
      mutationEvents.onNext(
        object : MutationEvent<E> {
          override fun state(): DirectedGraph<E> {
            return nextState
          }

          override fun operations(): Set<GraphOperation<E>> {
            return operations
          }
        })
    }
    return true
  }

  override fun nodeOf(item: E): DirectedGraph.DirectedNode<E>? {
    return nodeOfUnknownType(item)
  }

  override fun nodeOfUnknownType(item: Any?): DirectedGraph.DirectedNode<E>? {
    return getState().nodeOfUnknownType(item)
  }

  override fun edgeOf(from: E, to: E): DirectedGraph.DirectedEdge<E>? {
    return edgeOfUnknownType(from, to)
  }

  override fun edgeOfUnknownType(from: Any?, to: Any?): DirectedGraph.DirectedEdge<E>? {
    return getState().edgeOfUnknownType(from, to)
  }

  override fun contents(): Set<E> {
    return getState().contents()
  }

  override fun nodes(): Set<out DirectedGraph.DirectedNode<E>> {
    return getState().nodes()
  }

  override fun edges(): Set<out DirectedGraph.DirectedEdge<E>> {
    return getState().edges()
  }

  override fun toReadOnly(): ObservableDirectedGraph<E> {
    return object : ObservableDirectedGraph<E> {
      override fun observe(): ObservableDirectedGraph.ObservableChannels<E> {
        return this@WritableObservableDirectedGraphImpl.observe()
      }

      override fun nodeOf(item: E): DirectedGraph.DirectedNode<E>? {
        return nodeOfUnknownType(item)
      }

      override fun nodeOfUnknownType(item: Any?): DirectedGraph.DirectedNode<E>? {
        return this@WritableObservableDirectedGraphImpl.nodeOfUnknownType(item)
      }

      override fun edgeOf(from: E, to: E): DirectedGraph.DirectedEdge<E>? {
        return edgeOfUnknownType(from, to)
      }

      override fun edgeOfUnknownType(from: Any?, to: Any?): DirectedGraph.DirectedEdge<E>? {
        return this@WritableObservableDirectedGraphImpl.edgeOfUnknownType(from, to)
      }

      override fun nodes(): Set<out DirectedGraph.DirectedNode<E>> {
        return this@WritableObservableDirectedGraphImpl.nodes()
      }

      override fun edges(): Set<out DirectedGraph.DirectedEdge<E>> {
        return this@WritableObservableDirectedGraphImpl.edges()
      }

      override fun contents(): Set<E> {
        return this@WritableObservableDirectedGraphImpl.contents()
      }
    }
  }

  override fun observe(): ObservableDirectedGraph.ObservableChannels<E> {
    return object : ObservableDirectedGraph.ObservableChannels<E> {
      override fun states(): Observable<DirectedGraph<E>> {
        return Observable.create { emitter: ObservableEmitter<DirectedGraph<E>> ->
          synchronized(this) {
            emitter.onNext(getState())
            emitter.setDisposable(
              mutationEvents.map { obj: MutationEvent<E> -> obj.state() }
                .subscribe(
                    { emitter.onNext(it) },
                    { emitter.onError(it) },
                    { emitter.onComplete() }))
          }
        }
      }

      override fun mutations(): Observable<MutationEvent<E>> {
        return Observable.create { emitter: ObservableEmitter<MutationEvent<E>> ->
          synchronized(this) {
            val state = getState()
            val operations: Set<GraphOperation<E>> =
              Streams.concat(
                state.nodes().stream()
                  .map { it.item() }
                  .map { AddNodeToGraph.create(it) },
                state.edges().stream()
                  .map { it.endpoints() }
                  .map { couplet -> couplet.map(Function { it.item() }) }
                  .map { AddEdgeToGraph.create(it) })
                .collect(Collectors.toImmutableSet())
            emitter.onNext(object : MutationEvent<E> {
              override fun state(): DirectedGraph<E> {
                return state
              }

              override fun operations(): Set<GraphOperation<E>> {
                return operations
              }
            })
            emitter.setDisposable(
              mutationEvents
                .subscribe(
                    { emitter.onNext(it) },
                    { emitter.onError(it) },
                    { emitter.onComplete() }))
          }
        }
      }
    }
  }

  private fun getState(): ImmutableDirectedGraph<E> {
    synchronized(this) { return state }
  }

  internal interface MutationEvent<E : Any> : ObservableDirectedGraph.MutationEvent<E> {

    override fun state(): DirectedGraph<E>

    override fun operations(): Set<GraphOperation<E>>
  }

  private abstract class NodeOperation<E : Any>(private val item: E) :
      ObservableGraph.NodeOperation<E> {

    override fun item(): E {
      return item
    }
  }

  private class AddNodeToGraph<E : Any> private constructor(item: E) : NodeOperation<E>(item),
    ObservableGraph.AddNodeToGraph<E> {

    companion object {

      fun <E : Any> create(item: E): AddNodeToGraph<E> {
        return AddNodeToGraph(item)
      }
    }
  }

  private class RemoveNodeFromGraph<E : Any> private constructor(item: E) : NodeOperation<E>(item),
    ObservableGraph.RemoveNodeFromGraph<E> {

    companion object {

      fun <E : Any> create(item: E): RemoveNodeFromGraph<E> {
        return RemoveNodeFromGraph(item)
      }
    }
  }

  private abstract class EdgeOperation<E : Any>(private val endpoints: Couplet<E>) :
    ObservableGraph.EdgeOperation<E> {

    override fun endpoints(): Couplet<E> {
      return endpoints
    }
  }

  private class AddEdgeToGraph<E : Any> private constructor(endpoints: Couplet<E>) :
    EdgeOperation<E>(endpoints), ObservableGraph.AddEdgeToGraph<E> {

    companion object {

      fun <E : Any> create(endpoints: Couplet<E>): AddEdgeToGraph<E> {
        return AddEdgeToGraph(endpoints)
      }
    }
  }

  private class RemoveEdgeFromGraph<E : Any> private constructor(endpoints: Couplet<E>) :
    EdgeOperation<E>(endpoints), ObservableGraph.RemoveEdgeFromGraph<E> {

    companion object {

      fun <E : Any> create(endpoints: Couplet<E>): RemoveEdgeFromGraph<E> {
        return RemoveEdgeFromGraph(endpoints)
      }
    }
  }
}